ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+    int sg_start, sg_end;
+    int sg_size, log2_sg_size;
+    int subgrp_size, last_subgrp_size;
+    ptrdiff_t rlb, rext;
+    char *tmpsend = NULL, *tmprecv = NULL;
+    int sendto, recvfrom;
+    int num_data_blks;
+    size_t data_blk_size[2] = {0}, blk_ofst[2] = {0};
+    size_t bcount;
+    size_t last_subgrp_rcnt;
+    int brank, last_brank;
+    int use_rd_base, use_ring_sg;
+    int use_ring = 0, use_lin = 0;
+    int nreqs;
+    ompi_request_t **preq, **reqs;
+    size_t dsize;
+    mca_coll_acoll_module_t *acoll_module = (mca_coll_acoll_module_t *) module;
+
+    err = ompi_datatype_get_extent(rdtype, &rlb, &rext);
+    if (MPI_SUCCESS != err) {
+        return err;
+    }
+
+    ompi_datatype_type_size(rdtype, &dsize);
+    size = ompi_comm_size(comm);
+    rank = ompi_comm_rank(comm);
+    sg_size = acoll_module->sg_cnt;
+    log2_sg_size = acoll_module->log2_sg_cnt;
+
+    /* Handle non MPI_IN_PLACE */
+    tmprecv = (char *) rbuf + (ptrdiff_t) rank * (ptrdiff_t) rcount * rext;
+    if (MPI_IN_PLACE != sbuf) {
+        tmpsend = (char *) sbuf;
+        err = ompi_datatype_sndrcv(tmpsend, scount, sdtype, tmprecv, rcount, rdtype);
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+
+    /* Derive subgroup parameters */
+    sg_id = rank >> log2_sg_size;
+    num_sgs = (size + sg_size - 1) >> log2_sg_size;
+    sg_start = sg_id << log2_sg_size;
+    sg_end = sg_start + sg_size;
+    if (sg_end > size) {
+        sg_end = size;
+    }
+    subgrp_size = sg_end - sg_start;
+    last_subgrp_size = size - ((num_sgs - 1) << log2_sg_size);
+    last_subgrp_rcnt = rcount * last_subgrp_size;
+    use_ring_sg = (subgrp_size != sg_size) ? 1 : 0;
+    bcount = rcount << log2_sg_size;
+
+    /* Override subgroup params based on data size */
+    coll_allgather_decision_fixed(size, dsize * rcount, sg_size, &use_ring, &use_lin);
+
+    if (use_lin) {
+        err = ompi_coll_base_allgather_intra_basic_linear(sbuf, scount, sdtype, rbuf, rcount,
+                                                          rdtype, comm, module);
+        return err;
+    }
+    if (use_ring) {
+        sg_size = sg_end = subgrp_size = size;
+        num_sgs = 1;
+        use_ring_sg = 1;
+        sg_start = 0;
+    }
+
+    /* Do ring/recursive doubling based allgather within subgroup */
+    adj_rank = rank - sg_start;
+    if (use_ring_sg) {
+        recvfrom = ((adj_rank - 1 + subgrp_size) % subgrp_size) + sg_start;
+        sendto = ((adj_rank + 1) % subgrp_size) + sg_start;
+
+        /* Loop over ranks in subgroup */
+        for (i = 0; i < (subgrp_size - 1); i++) {
+            int recv_peer = ((adj_rank - i - 1 + subgrp_size) % subgrp_size) + sg_start;
+            int send_peer = ((adj_rank - i + subgrp_size) % subgrp_size) + sg_start;
+
+            tmprecv = (char *) rbuf + (ptrdiff_t) recv_peer * (ptrdiff_t) rcount * rext;
+            tmpsend = (char *) rbuf + (ptrdiff_t) send_peer * (ptrdiff_t) rcount * rext;
```

Comment:
Could you explain why tmpsend is using rbuf?

Reply:
Here we are using the ring algorithm where at each step, the rank sends the data it received in the previous step. Since `rbuf `contains the data received by the rank, it is used to derive `tmpsend`.